src/ml/neural_net/tf_activity_classifier_backend.cpp

```diff
@@ -0,0 +1,195 @@
+    PyGILState_STATE gstate;
+  	gstate = PyGILState_Ensure();
+
+    turi::scoped_finally gstate_restore([&](){PyGILState_Release(gstate);});
+
+	try {
+		func();
+	} catch(...){
+	    	log_and_throw("An error occurred!");
+	    }
+}
+
+
+static std::vector<size_t> get_dimensions( const float_array& num) {
+  std::vector<size_t> result(num.dim());
+  std::memcpy(result.data(), num.shape(), num.dim()* sizeof(size_t));
+  return result;
+}
+
+static std::vector<size_t> get_strides( const float_array& num) {
+  std::vector<size_t> result(num.dim());
+  std::fill(result.begin(), result.end(), sizeof(float));
+  return result;
+}
+
+PYBIND11_MODULE(libtctensorflow, m) {
+	py::class_<float_array>(m, "FloatArray", py::buffer_protocol())
+	.def_buffer([](float_array &m) -> py::buffer_info{
+	      return py::buffer_info(
+	          const_cast<float*>(m.data()),                   /* Pointer to buffer */
+	          sizeof(float),                                  /* Size of one scalar */
+	          py::format_descriptor<float>::format(),         /* Python struct-style format descriptor */
+	          m.dim(),                                        /* Number of dimensions */
+	          get_dimensions(m),                              /* Buffer dimensions */
+	          get_strides(m)                                  /* Strides (in bytes) for each index */
+	           
+	      );
+	});
+	py::class_<shared_float_array>(m, "SharedFloatArray", py::buffer_protocol())
+	.def_buffer([](shared_float_array &m) -> py::buffer_info{
+	      return py::buffer_info(
+	          const_cast<float*>(m.data()),                   /* Pointer to buffer */
+	          sizeof(float),                                  /* Size of one scalar */
+	          py::format_descriptor<float>::format(),         /* Python struct-style format descriptor */
+	          m.dim(),                                        /* Number of dimensions */
+	          get_dimensions(m),                              /* Buffer dimensions */
+	          get_strides(m)                                  /* Strides (in bytes) for each index */
+	            
+	      );
+	});
+}
+
+
+tf_activity_classifier_backend::tf_activity_classifier_backend(int batch_size, int num_features, 
+	int num_classes, int predictions_in_chunk, const float_array_map& config, const float_array_map& weights) {
+
+  	shared_float_array prediction_window = config.at("ac_pred_window");
+  	const float * pred_window = prediction_window.data();
+  	int pw = static_cast<int>(*pred_window);
+
+    call_pybind_function([&]() {	
+  	
+	py::module tf_ac_backend = py::module::import("turicreate.toolkits.activity_classifier._tf_model_architecture");
+	
+	// Make an instance of python object 
+	activity_classifier_ = tf_ac_backend.attr("ActivityTensorFlowModel")(weights, 
+		batch_size, num_features, num_classes, pw, predictions_in_chunk);
+
+    });
+
+}
+
+
+float_array_map tf_activity_classifier_backend::train(const float_array_map& inputs) {
```

Comment:
I don't see anything specific to AC in these method implementations. Could we factor it out into a C++ base class for all TF models?

Reply:
All the methods except the model object right?

Reply:
Yeah, I think the things to factor out all the implementations of the four methods in the `model_backend` interface